Documentation/teaching/lectures/syscalls.rst

```diff
@@ -0,0 +1,596 @@
+
+
+Since system calls are executed in kernel mode, they have access to
+kernel space and if pointers are not properly checked user
+applications might get read or write access to kernel space.
+
+For example, lets consider the case where such a check is not made for
+the read or write system calls. If the user passes a kernel-space
+pointer to a write sysytem call then it can get access to kernel data
+by later reading the file. If it passes an kernel-space pointer to a
+read system call then it can corrupt kernel memory.
+
+
+.. slide:: Pointers to Kernel Space
+   :level: 2
+
+   * User access to kernel data if allowed in a write system call
+
+   * User corrupting kernel data if allowed in a read system call
+
+
+Likewise, if a pointer passed by the application is invalid
+(e.g. unmaped, read-only for cases where its need to be written), it
+could "crash" the kernel. There two approaches that could be used:
+
+.. slide:: Invalid pointers
+   :inline-contents: True
+   :level: 2
+
+   * Check the pointer against the user address space
+
+   * Rely on the MMU to avoid checking the pointer for the case when
+     it is correct and handle the fault for the invalid case
+
+Although it sounds tempting, the second approach is not that easy to
+implement. The page fault handler uses the fault and faulting
+addresses as well as information from the user address space to
+determine the cause:
+
+.. slide:: Page fault handling
+   :inline-contents: True
+   :level: 2
+
+      * Copy on write, demand paging, swapping: both the fault and
+	faulting addresses are in userspace; the fault address is
+	valid (checked against the user address space)
+
+      * Invalid syscall pointer: the faulting address is in kernel
+	space; the fault address is in userspace and it is invalid
+
+      * Kernel bug: same as above
+
+But in the last two cases we don't have enough information to
+determine the cause of the fault.
+
+In order to solve this issue Linux uses special APIs (e.g
+:c:func:`copy_to_user`) to accesses userspace that are specially
+crafted:
+
+.. slide:: Marking kernel code that accesses user space
+   :inline-contents: True
+   :level: 2
+
+   * The exact instructions that accesses user space are recorded in
+     a table (exception table)
+
+   * When a page fault occurs the faulting address is checked against
+     this table
+
+
+Although the fault handling case may be more costly overall depending
+on the address space vs exception table size, and it is more complex,
+it does optimizes for the common case and that is why it is preferred.
+
+
+.. slide:: Cost analysis for pointer checks vs fault handling
+   :inline-contents: True
+   :level: 2
+
+   +------------------+-----------------------+------------------------+
+   | Cost             |  Pointer checks       | Fault handling         |
+   +==================+=======================+========================+
+   | Valid address    | address space search  | 0                      |
```

Comment:
[valid address - fault handling] I think it is O(1) because we still need to use copy_from_user which does some limits checking.